.repo/sample-list-generator/src/libs/api-detector.ts

```diff
@@ -0,0 +1,156 @@
+import {
+  ApiItem,
+  ApiItemWithType,
+  ApiTypeResult,
+  ExtensionApiMap
+} from '../types';
+import * as babel from '@babel/core';
+import { isMemberExpression } from '@babel/types';
+import fs from 'fs/promises';
+import { getAllFiles } from '../utils/filesystem';
+import { loadExtensionApis } from './api-loader';
+
+let EXTENSION_API_MAP: ExtensionApiMap = loadExtensionApis();
+
+export const getApiType = (
+  namespace: string,
+  propertyName: string
+): ApiTypeResult => {
+  namespace = namespace.replace(/_/g, '.');
+
+  if (EXTENSION_API_MAP[namespace]) {
+    const apiTypes = EXTENSION_API_MAP[namespace];
+
+    if (apiTypes.methods.includes(propertyName)) {
+      return 'method';
+    }
+    if (apiTypes.events.includes(propertyName)) {
+      return 'event';
+    }
+    if (apiTypes.properties.includes(propertyName)) {
+      return 'property';
+    }
+    if (apiTypes.types.includes(propertyName)) {
+      return 'type';
+    }
+  }
+  console.log('api not found', namespace, propertyName);
+  return 'unknown';
+};
+
+export const getApiListForSample = async (
+  sampleFolderPath: string
+): Promise<ApiItemWithType[]> => {
+  // get all js files in the folder
+  const jsFiles = (await getAllFiles(sampleFolderPath)).filter((file) =>
+    file.endsWith('.js')
+  );
+
+  const calls: ApiItemWithType[] = [];
+
+  const parallelHandler = jsFiles.map(async (file) => {
+    const _calls = await extractApiCalls(await fs.readFile(file));
+    calls.push(..._calls);
+  });
+  await Promise.all(parallelHandler);
+
+  return uniqueItems(calls);
+};
+
+export function getFullMemberExpression(
+  path: babel.NodePath<babel.types.MemberExpression>
+): string[] {
+  const result: string[] = [];
+  const property = path.node.property;
+
+  if (property.type === 'Identifier') {
+    result.push(property.name);
+  }
+
+  const parentNode = path.parentPath.node;
+
+  if (isMemberExpression(parentNode)) {
+    result.push(
+      ...getFullMemberExpression(
+        path.parentPath as babel.NodePath<babel.types.MemberExpression>
+      )
+    );
+  }
+
+  return result;
+}
+
+export function getApiItem(parts: string[]): ApiItem {
+  // special case for `chrome.storage`
+  if (parts[0] === 'storage') {
```

Comment:
If we get rid of this, what happens? I'm assuming it means we will no longer get "storage.onChanged" for "storage.local.onChanged" but I'm not sure how important that is and I feel like getting rid of the special case would be nice.

Reply:
It seems that there is indeed no such usage method in the document. I will remove this special case.

Reply:
Oh, I've thought about it and maybe this special case shouldn't be removed: 
`storage.sync` should correspond to `/reference/storage/#property-sync`, but `storage.sync.onChange` should correspond to `/reference/storage/#event-onChanged`, which is different from other APIs. However, my code can handle the situation of `storage.onChange`.

Reply:
Hmm. Shouldn't it actually link to https://developer.chrome.com/docs/extensions/reference/storage/#event-StorageArea-onChanged?

Reply:
Oh sorry, this is my mistake. I overlooked this. 😥